src/reducers/math.js

```diff
@@ -14,15 +23,75 @@ const mathReducer = produce((draft, action) => {
       setWith(
         draft.raw,
         type.replace(/^cdp/, 'cdps').replace(/^ilk/, 'ilks'),
-        value,
+        convertValue(type, value),
         Object
       );
 
-      // TODO: this is where we could notice changes in raw values that would
-      // cause changes in expensive-to-calculate derived values, and update
-      // their values as well
+      // here we can monitor changes to raw values that would affect
+      // expensive-to-calculate derived values, and update their values as well
+
+      if (type.startsWith('ilk')) {
+        const [, name, prop] = type.split('.');
+
+        // update the price for an ilk if `spot` or `mat` changes
+        if ([PRICE_WITH_SAFETY_MARGIN, LIQUIDATION_RATIO].includes(prop)) {
+          const feed = draft.feeds.find(f => f.key === name);
+          feed[FEED_VALUE_USD] = recalculatePrice(
+            draft.raw.ilks[name],
+            name,
+            draft.system[PAR]
+          );
+        }
+      }
+
+      // if `par` changes (which is unlikely) all the prices need to change
+      if (type === `system.${PAR}` && draft.raw.ilks) {
```

Comment:
It seems to make sense to me to have this actually belong in the feeds reducer, and have it listen for the `par` since we are caching the information we need to calculate the price. What's the motivation behind using the raw data directly?

Reply:
The mental model I'm aiming toward is basically: "store only raw values, and calculate derived values on the fly unless deriving them is expensive." So I imagine the current feeds reducer being used less and less, as we use the [dai-plugin-mcd math functions](https://github.com/makerdao/dai.js/blob/dev/lib/dai-plugin-mcd/src/math.js) which work with raw values more and more. And we would [store derived values](https://github.com/makerdao/mcd-cdp-portal/pull/72/files#diff-f1559a7189a2a74c87c358c05a99b25fR63) only in cases where the raw value is never used for calculations.

I believe this model is valuable because it helps us encapsulate all the conversion and calculation into one place, so that utility functions and components can have simple expectations for their input--basically "it's a BigNumber or it's a string that I have to do nothing with except render". Right now this logic is in utils/ui, utils/bignumber, utils/cdps, reducers/cdps, et al., and I feel the mental overhead of having to check if each of these functions takes a Number, BigNumber, string, or an object composed of such values.

In my imagined end state, there isn't really a need to have separate feeds, system, and cdps reducers, since the core reducer work becomes just [setWith](https://github.com/makerdao/mcd-cdp-portal/pull/72/files#diff-f1559a7189a2a74c87c358c05a99b25fR23).

But maybe this is too much of a departure from a structure that other people are already comfortable with? Let me know.

Reply:
The mental model I'm aiming toward is basically: "store only raw values, and calculate derived values on the fly unless deriving them is expensive." So I imagine the current feeds reducer being used less and less, as we use the [dai-plugin-mcd math functions](https://github.com/makerdao/dai.js/blob/dev/lib/dai-plugin-mcd/src/math.js) which work with raw values more and more. And we would [store derived values](https://github.com/makerdao/mcd-cdp-portal/pull/72/files#diff-f1559a7189a2a74c87c358c05a99b25fR63) only in cases where the raw value is never used for calculations.

I believe this model is valuable because it helps us encapsulate all the conversion and calculation into one place, so that utility functions and components can have simple expectations for their input--basically "it's a BigNumber or it's a string that I have to do nothing with except render". Right now this logic is in utils/ui, utils/bignumber, utils/cdps, reducers/cdps, et al., and I feel the mental overhead of having to check if each of these functions takes a Number, BigNumber, string, or an object composed of such values.

In my imagined end state, there isn't really a need to have separate feeds, system, and cdps reducers, since the core reducer work becomes just [setWith](https://github.com/makerdao/mcd-cdp-portal/pull/72/files#diff-f1559a7189a2a74c87c358c05a99b25fR23).

But maybe this is too much of a departure from a structure that other people are already comfortable with? Let me know.

/cc @jparklev @michaelelliot

Reply:
I'm going to merge this now, but let's keep the discussion going. Always open to changing it down the road if we reach a different consensus. It's never set in stone.

Reply:
> Right now this logic is in utils/ui, utils/bignumber, utils/cdps, reducers/cdps, et al., and I feel the mental overhead of having to check if each of these functions takes and/or returns a number, BigNumber, string, or an object composed of such values.

+100

I really like this model (and the way `setWith` is used makes sense to me). So, rather than a "math" reducer, it becomes a kind of general "chain state" reducer/ cache? One thing I notice with this is the ways in which it approaches how multicall works internally; ie multicall keeps a object like `{ [<event_type>]: <value>, ... }` to, after the initial fetch, check if the state returned from the node for each event type has changed. I wonder about how these things might converge in the future.

Reply:
Cool. I named it "math" to gesture at the idea that it would also store derived values, but it's really just a placeholder name... Interesting to know that's what multicall does.